[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56504 [PyTorch] Replace RecordFunction shouldRun callback with atomic bools
* #56492 [PyTorch] Remove dead get/setTLSCallbacks APIs
* **#56115 [PyTorch] Use MaybeOwned and avoid resize in bmm_cuda**

No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

swolchok committed:
[PyTorch] Allocate correctly-sized output tensor in bmm_cuda

No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok commented on aten/src/ATen/native/cuda/LinearAlgebra.cu:
```diff
@@ -195,7 +195,7 @@ Tensor& baddbmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor&
   TORCH_CHECK(self_sizes[0] == batch2_sizes[0], "self dim 0 must match batch2 dim 0");
   TORCH_CHECK(self_sizes[1] == batch1_sizes[1], "self dim 1 must match batch1 dim 1");
-  TORCH_CHECK(self_sizes[2] == batch2_sizes[2], "self dim 2 must match batch2 dim 2");
+  TORCH_CHECK(self_sizes[2] == batch2_sizes[2], "self dim  2 must match batch2 dim 2");
```
oops

IvanYashchuk commented:
Windows build tests fail with `CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling cublasZgemmStridedBatched`. cuBLAS documentation says that it means the parameters `m,n,k,batchCount<0`.

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
fix build at bottom of stack on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use MaybeOwned and avoid resize in bmm_cuda"


No reason to size it wrong and then resize it. Also, no reason to unconditionally go through the dispatcher.

Differential Revision: [D27768757](https://our.internmc.facebook.com/intern/diff/D27768757/)

[ghstack-poisoned]
